Add runtime for AWS Lambda

jleedev opened the pull request:
This adds the lambda-web crate to adapt the actix App to speak to Lambda by way of the lambda_runtime crate.

Briefly: AWS Lambda has native support for scripting languages to execute a function directly; compiled languages must embed a runtime to fetch incoming events from Lambda and post the responses. This detects the environment variables to start up in Lambda mode instead of the normal HTTP server, and is added as an optional feature.

Lambda has five (!) distinct ways of routing HTTP requests to a function; this supports some of them. (Specifically, the most obvious way to do this is with a Function URL, which is newest and simplest, and perhaps with CloudFront, which speaks to the Function URL and not Lambda directly.)

The error handling could probably be refined, I was just trying to get this to compile.

(Supported: API Gateway HTTP API with payload format version 2.0; API Gateway REST API; Lambda function URLs / Not supported: API Gateway HTTP API with payload format version 1.0; Application Load Balancer)

Necessary for #1102 to be able to run the released packages directly, and only having to configure the appropriate environment.

### TODOs

* [x] Update Book Documentation
* [ ] Update README
* [x] Add CI test(s)

jleedev committed:
Add runtime for AWS Lambda

nyurik commented on martin/Cargo.toml:
```diff
@@ -60,8 +60,9 @@ name = "bench"
 
 [features]
-default = ["fonts", "mbtiles", "pmtiles", "postgres", "sprites"]
+default = ["fonts", "lambda", "mbtiles", "pmtiles", "postgres", "sprites"]
```
do we want to group all lambda supports into one group, or will we need to use different dependencies/code-paths for aws vs gcp vs azure?

jleedev replied:
Fortunately, only Lambda (to my knowledge) requires shenanigans like this, everything else supports regular HTTP servers.

nyurik commented on martin/src/srv/server.rs:
```diff
@@ -127,15 +127,32 @@ pub fn new_server(config: SrvConfig, state: ServerState) -> MartinResult<(Server
+
+    match () {
+        #[cfg(feature = "lambda")]
+        () if is_running_on_lambda() => {
```
why not a simple block like this?  Why do you need a match statement?

```
#[cfg(feature = "lambda")]
if is_running_on_lambda() {
    ....
}
```

nyurik replied:
P.S. we need to document that a new env var variable has magical properties per https://github.com/hanabu/lambda-web/blob/ddceea3b8ffcfae465fc8d0059aaa52a8ff94f8e/src/lib.rs#L34 

if `AWS_LAMBDA_RUNTIME_API` is defined, runs under AWS lambda runtime

jleedev replied:
I think I was frustrated that an if/else wouldn't work with the conditional compilation, and this is a way to express that only one of the code paths will run. But an if statement and a return would probably be clearer.

nyurik replied:
I may refactor it later, no worries. I think we should focus on the main issue - Testing, and ideally - testing in CI.  We need a clear way to ensure that after some person adds some random code, we don't have to do too much work to make sure it continues to work in all expected scenarios.

nyurik reviewed (commented):
looks great!  My biggest concern is how can we test this, especially as part of the CI?

jleedev committed:
Flatten match

nyurik commented on martin/src/bin/martin.rs:
```diff
@@ -9,7 +9,7 @@ use martin::{read_config, Config, MartinResult};
 const VERSION: &str = env!("CARGO_PKG_VERSION");
 
-async fn start(args: Args) -> MartinResult<Server> {
+async fn start(args: Args) -> MartinResult<impl Future<Output = MartinResult<()>>> {
```
this looks ... scary. Could you add a comment explaining why you have future of a future here, or if we can somehow simplify this?  Basically here we have a desugared `Future<Result<Future<Result>>>>`, right?

jleedev replied:
Exactly, the actix `Server` type was just being used as a future, and desugars to this. I note that the main function is already awkward with the `.await.unwrap().await.unwrap()`.

The only reason to do anything with the server besides awaiting it immediately would be to pause or stop it, and also to print out the listening address as we do here.

Defining a type alias for what the "Server" is would reduce the clutter in the other file, but this function can really be changed to return `MartinResult<()>`, since this is essentially the main.

nyurik commented on martin/src/srv/server.rs:
```diff
@@ -98,16 +100,14 @@ pub fn router(cfg: &mut web::ServiceConfig) {
+pub fn new_server(
+    config: SrvConfig,
+    state: ServerState,
+) -> MartinResult<(LocalBoxFuture<'static, MartinResult<()>>, String)> {
```
same thing as above - any way to simplify the result of this, or at least lots of comments?

nyurik reviewed (commented):
much simpler, thanks!  I added a few more comments, but this is looking better and better, almost ready to go.  I will add a few TODOs to the top comment.

jleedev committed:
inline await in main; add type alias

jleedev committed:
Merge remote-tracking branch 'origin/main' into lambda

jleedev committed:
Add a Lambda smoke test

Verifies that the (amd64) container image can start up under AWS Lambda
and respond to an HTTP event with a 200. Does not supply any
configuration or test any handlers.

jleedev committed:
appease the clippy

jleedev committed:
pass event correctly

jleedev committed:
use jq properly - error on empty input

jleedev committed:
Move lambda test to script file

jleedev committed:
syntax error

jleedev committed:
Add documentation for AWS Lambda

jleedev committed:
Make the Lambda test pass

Switch it from image to zip to make it easier to load in a pmtiles file

jleedev committed:
just fmt-md

jleedev committed:
yaml syntax

jleedev committed:
workflow syntax

jleedev committed:
cleanup

jleedev committed:
Describe AWS_LAMBDA_RUNTIME_API env var

jleedev committed:
typo

jleedev committed:
Switch from gnu to musl for the Lambda CI

On the hopes that al2023 won't complain about:

    martin: /lib64/libm.so.6: version `GLIBC_2.35' not found (required by martin)

jleedev committed:
yaml typo

jleedev committed:
debugging this artifact nonsense

jleedev committed:
correct path

jleedev committed:
oh lol it unzips it for you

jleedev committed:
typo

jleedev committed:
correct chmod

nyurik commented on .github/files/lambda-function/Makefile:
```diff
@@ -0,0 +1,2 @@
+build-MartinFunction:
```
where is this makefile being used?

jleedev replied:
Implicitly by `sam build` of the AWS::Serverless::Function. It _requires_ a Makefile. (Or like, something such as Java or JavaScript or Python which it knows how to build.)

(But AWS::Serverless::LayerVersion is happy to just zip up a directory. Which is a little bit silly since that's the complicated one and the function code is just one yaml, but oh well.)

See: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-custom-runtimes.html and https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-rust.html

Note that they also have a `cargo lambda` utility which is in beta, but might be worth a look later.

nyurik commented on .github/files/lambda-layer/bootstrap:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+set -eu
+exec martin -c "${_HANDLER}"
```
will `martin` be available in the PATH?  I guess it is if it passes tests?

jleedev replied:
Yes, because it's under `bin/` in the zip file, and the layer extracts somewhere on Lambda's path.

nyurik commented on docs/src/run-with-lambda.md:
````diff
@@ -0,0 +1,124 @@
+tar -C src/bin/ -xzf martin-aarch64-unknown-linux-musl.tar.gz martin
+```
+
+Every zip-based Lambda function runs a file called `bootstrap`. `vim src/bootstrap`
````
better use heredoc for this - easier, and does not require familiarity with `vim`.  Might also use `#!/usr/bin/env sh`, but not certain which shebang is better.

```
cat <<EOF > src/bootstrap
#!/bin/sh
set -eu
exec martin -c ${_HANDLER}.yaml
EOF
```

jleedev replied:
Looks like AL2023's /bin/sh is "GNU bash, version 5.2.15(1)-release (x86_64-amazon-linux-gnu)", and they also have a merged / and /usr. So it should be fine.

I'm amused to learn that `#!/bin/sh` is actually not a POSIX compliant thing to do, as the path to `sh` cannot be assumed.

Anyway, `/bin/sh` causes Bash to start up in its posix mode, and refers to Dash on Debian and Ubuntu systems, but this small shell script shouldn't encounter any differences :)

nyurik commented on docs/src/run-with-lambda.md:
````diff
@@ -0,0 +1,124 @@
+exec martin -c ${_HANDLER}.yaml
+```
+
+Write the SAM template. `vim template.yaml`
````
use heredoc as above

nyurik commented on docs/src/run-with-lambda.md:
```diff
@@ -0,0 +1,124 @@
+
+2. Click Deploy, wait for the success banner, and visit your function URL.
+
+### Extra things
```
"things" here looks weird - either `Extras` or `TODO` ?

```suggestion
### TODO
```

nyurik commented on docs/src/env-vars.md:
```diff
@@ -9,3 +9,4 @@ You can also configure Martin using environment variables, but only if the confi
 | `PGSSLCERT` <br/> `ssl_cert`             | `./postgresql.crt`                   | A file with a client SSL certificate. [docs](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-SSLCERT)                                                                             |
 | `PGSSLKEY` <br/> `ssl_key`               | `./postgresql.key`                   | A file with the key for the client SSL certificate. [docs](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-SSLKEY)                                                                |
 | `PGSSLROOTCERT` <br/> `ssl_root_cert`    | `./root.crt`                         | A file with trusted root certificate(s). The file should contain a sequence of PEM-formatted CA certificates. [docs](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-SSLROOTCERT) |
+| `AWS_LAMBDA_RUNTIME_API`                 |                                      | If defined, connect to AWS Lambda to handle requests. The regular HTTP server is not used.                                                                                                                         |
```
This should probably link to the book's page on AWS lambda

nyurik commented on docs/src/run-with-lambda.md:
```diff
@@ -0,0 +1,124 @@
+6. Set “Architecture” to arm64 to match the platform that we pulled. (Lambda has better ARM CPUs than x86.)
+7. Click “Create function”.
+8. Find the “Configuration” tab, select “Function URL”, “Create function URL”.
+9. Set “Auth type” to NONE. (Don’t enable CORS, as Martin already will, and you will have duplicate headers and your CORS will be broken.)
```
```suggestion
9. Set “Auth type” to `NONE`
   * Do not enable CORS. Martin already has CORS support, so it will create duplicate headers and break CORS.
```

nyurik commented on docs/src/run-with-lambda.md:
````diff
@@ -0,0 +1,124 @@
+```bash
+mkdir -p martin_layer/src/bin/
+cd martin_layer
+curl -OL https://github.com/maplibre/martin/releases/download/VERSION_NUMBER_HERE/martin-aarch64-unknown-linux-musl.tar.gz
````
Use `latest` instead of requiring version number

```suggestion
curl -OL https://github.com/maplibre/martin/releases/latest/download/martin-aarch64-unknown-linux-musl.tar.gz
```

nyurik commented on docs/src/run-with-lambda.md:
```diff
@@ -0,0 +1,124 @@
+
+Run `sam deploy --guided`.
+
+1. Stack Name: [martin-layer]
```
broken link?

nyurik commented on docs/src/run-with-lambda.md:
```diff
@@ -0,0 +1,124 @@
+
+Add your layer:
+
+1. Click “add a layer” (green banner at the top, or the very bottom).
```
in cases like these, you may want to add screenshot images, but we can do this in a separate PR.

nyurik commented on docs/src/run-with-lambda.md:
```diff
@@ -0,0 +1,124 @@
+### Extra things
+
+- Lambda has a default timeout of 3 seconds, and 128 MB of memory, maybe this is suboptimal.
+- Martin will cache tiles in memory, but this might be useless for an ephemeral server.
```
I don't think this is useless - if martin binary will handle more than one request, it will be better to cache at least some things - for example, when using pmtiles, cache the directory structure.

nyurik commented on martin/src/bin/martin.rs:
```diff
@@ -35,20 +34,15 @@ async fn start(args: Args) -> MartinResult<Server> {
-        .unwrap_or_else(|e| on_error(e))
-        .await
-        .unwrap_or_else(|e| on_error(e));
+    start(Args::parse()).await.unwrap_or_else(|e| on_error(e));
```
probably better to inline the on_error here. Also, I noticed the same issue in `martin-cp` if you want to inline it there at the same time.

```suggestion
    start(Args::parse()).await.unwrap_or_else(|e| {
        if log_enabled!(log::Level::Error) {
            error!("{e}");
        } else {
            eprintln!("{e}");
        }
        std::process::exit(1);
    });
```

nyurik commented on tests/test-aws-lambda.sh:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
were you able to run this script locally?  If so, could you add a new recipe to the `justfile` too, showing how it can be run/tested locally?  thx

jleedev replied:
Ooh, even better:

```bash
 ¶ git diff .github/files/lambda.yaml
@@ -17,3 +17,6 @@ Resources:
         - Ref: MartinLayer
       CodeUri: lambda-function/
       Handler: config.yaml
+      Events:
+        ApiEvent:
+          Type: HttpApi
 ¶ sam local start-api --warm-containers eager
```

```bash
curl http://127.0.0.1:3000
```

Function URL is the best choice for running on AWS, but for local emulation we can configure an API Gateway HTTP API, and then send it HTTP requests instead of having to use `aws lambda invoke` with specially crafted event objects. Good for a followup.

nyurik replied:
@jleedev would it make sense to create an issue to track it? Won't remember otherwise.

nyurik reviewed (commented):
awesome!!!  i think its almost done. I added a few nits that should be very easy to fix, and just to clarify for myself if they are needed and how they are being used.  Thanks again!

jleedev committed:
doc changes

jleedev committed:
doc update

jleedev committed:
usage

jleedev committed:
inline the on_error

jleedev committed:
Add test-lambda to justfile and some usage notes

jleedev committed:
Add link

jleedev committed:
Update docs/src/run-with-lambda.md

Co-authored-by: Yuri Astrakhan <yuriastrakhan@gmail.com>

jleedev committed:
Update docs/src/run-with-lambda.md

Co-authored-by: Yuri Astrakhan <yuriastrakhan@gmail.com>

jleedev committed:
just fmt-md

nyurik reviewed (approved):
awesome work, thank you!!!

We may want to create follow up issues to remind later of any further tasks, but this is totally up to you :)
